docs/design/services/s3/transfermanager/prototype.java

```diff
@@ -0,0 +1,753 @@
+                        .presignedUrl(presignedUrl)
+                        .build(),
+                file);
+    }
+
+    /**
+     * UploadObject a file to S3.
+     */
+    default UploadObject uploadObject(UploadObjectRequest request, Path file) {
+        return uploadObject(request, ObjectPartStreamCreator.forFile(file));
+    }
+
+    /**
+     * UploadObject an object to S3, using {@code partStreamFactory} to provide
+     * stream(s) for its part(s).
+     * <p>
+     * This is intended for advanced use not covered by other overloads of
+     * {@code uploadObject} as implementing a {@code Publisher<ByteBuffer>} can
+     * be difficult and error prone. When possible, prefer the other overloads
+     * of {@code downloadObject}, or implementations of {@link
+     * ObjectPartStreamCreator} provided by the SDK.
+     * <p>
+     * <b>Usage Example</b>
+     * <pre>
+     * UploadObject myUpload = tm.uploadObject(UploadObjectRequest
+     *         .builder()
+     *         .presignedUrl(url)
+     *         .build(),
+     *         ObjectPartStreamFactory.forInputStream(myStream));
+     * myUpload.completionFuture().join();
+     * </pre>
+     *
+     * @param request The upload request.
+     * @param partStreamFactory The part stream factory.
+     */
+    UploadObject uploadObject(UploadObjectRequest request, ObjectPartStreamCreator partStreamFactory);
+
+    /**
+     * Resume a previously paused object upload.
+     */
+    UploadObject resumeUploadObject(UploadObjectState uploadObjectState);
+
+    /**
+     * Create an {@code S3TransferManager} using the default values.
+     */
+    static S3TransferManager create() {
+        return builder().build();
+    }
+
+    static S3TransferManager.Builder builder() {
+        return ...;
+    }
+
+    interface Builder {
+        /**
+         * The custom S3AsyncClient this transfer manager will use to make calls
+         * to S3.
+         */
+        Builder s3client(S3AsyncClient s3Client);
+
+        /**
+         * The configuration for this transfer manager.
+         */
+        Builder configuration(TransferManagerConfiguration configuration);
+
+        /**
+         * Add a progress listener to the currently configured list of
+         * listeners.
+         */
+        Builder addProgressListener(TransferProgressListener progressListener);
+
+        /**
+         * Set the list of progress listeners, overwriting any currently
+         * configured list.
+         */
+        Builder progressListeners(Collection<? extends TransferProgressListener> progressListeners);
+
+        S3TransferManager build();
+    }
+}
+
+/**
+ * Configuration object for {@link S3TransferManager}.
+ */
+public interface TransferManagerConfiguration {
+    /**
+     * The max number of requests the Transfer Manager will have at any point in
+     * time. This must be less than or equal to the max concurrent setting on
+     * the S3 client.
+     */
+    public int maximumConcurrency();
```

Comment:
Does this method name match the one in the netty client?

Reply:
Yup https://github.com/aws/aws-sdk-java-v2/blob/fbec1bf7ba835d7f98d073828d8954c87c2e30bc/http-clients/netty-nio-client/src/main/java/software/amazon/awssdk/http/nio/netty/NettyNioAsyncHttpClient.java#L448

Reply:
Nit: `maxConcurrency` vs `maximumConcurrency`

Reply:
Woops good catch